### PR TITLE
Updated TypeScript version into 2.4.1

### PR DESCRIPTION
**Pull request by aleixmorgadas**
set noStrictGenericChecks typescript compiler option to true due to Generic Error

### CODE REVIEW

**Comment by aleixmorgadas** [comment]
The fix implied more changes than expected:

- [X] Upgraded TypeScript version into 2.4.1
- [X] Upgraded TypeScript lib from es5 to es6 due type errors
- [X] Added @types/core-js dependency
- [X] Modify all imports to pass the tslint
- [X] string-literals.ts convertType function, replace type.text by <string> type.value

**Comment by aleixmorgadas** [comment]
- [X] Update gruntfile.js to add the compiler options
- [X] Change react.d.ts for known FIX [link](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/17578)

**Comment by aleixmorgadas** [comment]
Finally, the **build compiles** and run the tests with
- 61 passing (44s)
- **52 failing**

I need to do a lot of changes and fixes, so I think that I've done something wrong in some stage, could a core developer help me to fix this? Thanks

I leave the pull request at this stage to have a changes track

**Comment by drewwyatt** [comment]
Any idea when this will get released?

File: tsconfig.json [hunk 1]
```diff
@@ -16,7 +16,10 @@
     "outDir": "dist/",
     "rootDir": "src/",
     "experimentalDecorators": true,
-    "noStrictGenericChecks": true
+    "noStrictGenericChecks": true,
+    "typeRoots": [
```
Comment (aciccarello) [inline tsconfig.json]: This should not be necessary. TypeScript defaults to this.

File: src/test/converter/decorators/decorators.ts [hunk 2]
```diff
@@ -43,6 +43,6 @@ function decoratorWithParam(value:boolean):MethodDecorator {
  */
 function decoratorWithOptions(options:{name:string}):ClassDecorator {
     return function (target) {
-        target.options = options;
+        target.arguments = target;
```
Comment (aciccarello) [inline src/test/converter/decorators/decorators.ts]: So there error here is caused by a lib.es5.d.ts definition. Granted, this is only an example being documented but I'd prefer to update the definition of the target function to include this property.

File: src/index.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
-export {Application} from './lib/application';
-export {CliApplication} from './lib/cli';
+export {Application } from './lib/application';
+export {CliApplication } from './lib/cli';
```
Comment (aleixmorgadas) [inline src/index.ts]: I missed the file, strange that tslint didn't notice it

**Comment by aleixmorgadas** [comment]
The major changes made are to support TS 2.4 as soon as possible, and then remove the parches to apply the best practices. I'm OK with your decision to first apply the best procedure and then release the new version.

So, I will make a new PR for the tslint issue in the next days. How do you want to procedure with the ts version upgrade?

**Comment by aleixmorgadas** [comment]
👍

**Comment by abergs** [comment]
We're really looking forward to the release of this!

**Comment by aciccarello** [comment]
@aleixmorgadas Thanks for doing the whitespace changes in a separate PR. I made some changes to this PR to tighten the generics type definitions and avoid some of the CI timeouts.

@blakeembrey Can you take a quick look and create a release for this? The version number was already bumped.

**Comment by CosminIonascu** [comment]
Hi, I also need this update, when will be merged?

**Comment by abergs** [comment]
@CosminIonascu In the meantime you can install this pull request by running:

yarn add -D TypeStrong/typedoc#549/head
or
npm install --save-dev TypeStrong/typedoc#549/head

**Comment by CosminIonascu** [comment]
Ok, thank's!

**Comment by abergs** [comment]
Also @CosminIonascu, you will ned to build the dist folder in order be able to run the node script:
https://github.com/TypeStrong/typedoc/blob/master/UPDATING.md

File: package.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "typedoc",
   "description": "Create api documentations for typescript projects.",
-  "version": "0.7.1",
+  "version": "0.7.2",
```
Comment (blakeembrey) [inline package.json]: If this can be undone, I'll do a release.

File: tsconfig.json [hunk 5]
```diff
@@ -3,11 +3,9 @@
     "module": "commonjs",
     "lib": [
       "dom",
-      "es5",
-      "es2015.collection",
-      "es2015.iterable"
+      "es2016"
```
Comment (blakeembrey) [inline tsconfig.json]: `es2016`? Is there something from `es2016` we're using or should this just be `es2015`?

Comment (aleixmorgadas) [inline tsconfig.json]: there was some issues with the types due to the last typescript version, it was solved using the es2016 lib

File: package.json (hunk 4 above)
Comment (aleixmorgadas) [inline package.json]: Done 😁

File: tsconfig.json (hunk 5 above)
Comment (blakeembrey) [inline tsconfig.json]: Which issues?

**Comment by aleixmorgadas** [comment]
@blakeembrey I've replaced the lib es2016 by the old once to show you the errors.

So, the build now is broken just to show you what happens. If it's OK, then undo my last changes on tsconfig.json 😁

File: tsconfig.json (hunk 5 above)
Comment (aleixmorgadas) [inline tsconfig.json]: https://travis-ci.org/TypeStrong/typedoc/builds/257837771?utm_source=github_status&utm_medium=notification

**Comment by aleixmorgadas** [comment]
@blakeembrey I've undone the change in the `tsconfig.json` file, and it works.

There was a lot of errors at the beginning, so they guided me to that change. View all thread to find the reason.

**Comment by blakeembrey** [comment]
@aleixmorgadas The changes you made were invalid JSON. I'll try to check it out for myself.

**Comment by aleixmorgadas** [comment]
@blakeembrey Yes the commit d3d4d42 had an invalid JSON, but the first errors weren't.

**Comment by blakeembrey** [comment]
@aleixmorgadas As far as I can see, there's no changes to `tsconfig.json` in the PR I merged. It looks like you undid the changes making it `es2016` and it passed. Let me know if I'm not looking at the right thing.

**Comment by blakeembrey** [comment]
Oh, I see what you meant now too. I thought you were undoing the change to demonstrate it failing. The order of the conversation confused me, sorry. I saw your TravisCI link first and I thought that was you showing it fails using the `es2015` config. Now I understand that you undid the changes all together (not to `es2015` as a demo, then back to `es2016`).

**Pull request merged**